---
title: "MIT 20.440 Final Project: sncRNA transcriptomics as a predictive marker of Parkinson’s disease - all data analysis besides machine learning"
author: "Evan Collins and Brian Cheng"
date: "1 May 2023"
output: html_document
---

This R markdown document is best viewed and run in RStudio. Running each chunk will result in the production of all figures (besides those involving machine learning) as featured in our MIT 20.440 final project entitled *sncRNA transcriptomics as a predictive marker of Parkinson’s disease*.

# 0. Load necessary packages

```{r, warning = FALSE, message = FALSE}
# All R code written by Evan Collins for MIT 20.440 project
# All data acquired from Parkinson's Progression Markers Initiative (PPMI)
# 1 May 2023
# Any questions can be sent to evanc@mit.edu

# Install packages if necessary
install.packages("data.table",
                 version = "1.14.8",
                 repos = "https://cloud.r-project.org")
install.packages("umap", 
                 version = "0.2.10.0",
                 repos = "https://cloud.r-project.org")
install.packages("ggplot2",
                 version = "3.4.2",
                 repos = "https://cloud.r-project.org")
install.packages("ggprism", 
                 version = "1.0.4",
                 repos = "https://cloud.r-project.org")
install.packages("rstatix",
                 version = "0.7.2",
                 repos = "https://cloud.r-project.org")
install.packages("factoextra", 
                 version = "1.0.7",
                 repos = "https://cloud.r-project.org")
install.packages("MASS",
                 version = "7.3.58.3",
                 repos = "https://cloud.r-project.org")
install.packages("dplyr", 
                 version = "1.1.1",
                 repos = "https://cloud.r-project.org")
install.packages("ggpubr",
                 version = "0.6.0",
                 repos = "https://cloud.r-project.org")
install.packages("BiocManager",
                 version = "1.30.20",
                 repos = "https://cloud.r-project.org")
BiocManager::install("DESeq2")
BiocManager::install('EnhancedVolcano')

# Load packages
library(data.table)
library(umap)
library(ggplot2)
library(ggprism)
library(rstatix)
library(factoextra)
library(MASS)
library(dplyr)
library(ggpubr)
library(BiocManager)
library(EnhancedVolcano)
library(DESeq2)
```

# 0. Initial sncRNA and clinical data processing

(Due to data usage agreements, note that PPMI data is not included in the GitHub repo)

```{r}
# Load sncrna raw data
sncrna_df <- as.data.frame(fread("Data/PPMI_sncRNAcounts/counts/sncrna_quantification_matrix_raw.csv"))

# Get patients list
PATNO_list <- gsub("\\..*","", substring(colnames(sncrna_df)[-1], 10))
# Remove patient #221 (has NA in title;)
PATNO_list_clean <- PATNO_list[-221]
# Get unique patient ID numbers
PATNO_list_clean <- as.integer(PATNO_list_clean)
PATNO_unique_list <- unique(PATNO_list)
```

```{r}
# Load patient dictionary file included in PPMI database
participant_status_df <- read.csv("Data/Participant_Status.csv")
```

```{r}
# Further patient processing
# 335 of the patients are not in Participant_Status.csv
# Remove them
PATNO_list_clean <- PATNO_list[PATNO_list %in% participant_status_df$PATNO]
PATNO_unique_list_clean <- PATNO_unique_list[PATNO_unique_list %in% participant_status_df$PATNO]

sncrna_df_clean <- sncrna_df[, c(FALSE, PATNO_list %in% PATNO_list_clean)]
rownames(sncrna_df_clean) <- sncrna_df[,1]

# Get vector of cohort designations across all patients in sncRNA cohort
sncrna_cohort_full_list <- rep(NA, length(PATNO_list_clean))
for (i in 1:length(PATNO_list_clean)){
 sncrna_cohort_full_list[i] <- participant_status_df$COHORT_DEFINITION[participant_status_df$PATNO == PATNO_list_clean[i]]
}
sncrna_cohort_full_list[sncrna_cohort_full_list == "Healthy Control"] <- "Healthy"
sncrna_cohort_full_list[sncrna_cohort_full_list == "Parkinson's Disease"] <- "PD"

sncrna_cohort_list <- rep(NA, length(PATNO_unique_list_clean))
for (i in 1:length(PATNO_unique_list_clean)){
 sncrna_cohort_list[i] <- participant_status_df$COHORT_DEFINITION[participant_status_df$PATNO == PATNO_unique_list_clean[i]]
}
sncrna_cohort_list[sncrna_cohort_list == "Healthy Control"] <- "Healthy"
sncrna_cohort_list[sncrna_cohort_list == "Parkinson's Disease"] <- "PD"

# Make a patient info dictionary
patient_info_df <- data.frame("PATNO" = PATNO_unique_list_clean,
                              "cohort" = sncrna_cohort_list,
                              "num_visits" = NA)

for (i in 1:nrow(patient_info_df)){
  patient_info_df$num_visits[i] <- length(which(PATNO_list_clean %in% PATNO_unique_list_clean[i]))
}
```

```{r}
# Print out the number of baseline patients in each cohort
table(patient_info_df$cohort)
```

Note Figure 1 as featured in the write-up is a methods-based figure not rendered using R.


# Figure 2 - UMAP and LDA of all sncRNAs

```{r}
# UMAP
# Create baseline only sncRNA dataframe
sncrna_df_bl <- sncrna_df_clean[, grepl("BL", colnames(sncrna_df_clean))]
sncrna_df_bl_t <- as.data.frame(t(sncrna_df_bl)) # transpose such that rows are patients; columns are transcripts
PATNO_bl_list <- gsub("\\..*","", substring(rownames(sncrna_df_bl_t), 10))
# Make list of which cohort each patient belong to
cohort_bl_list <- c()
for (i in 1:length(PATNO_bl_list)){
  cohort_bl_list[i] <- participant_status_df$COHORT_DEFINITION[participant_status_df$PATNO == PATNO_bl_list[i]]
}
cohort_bl_list[cohort_bl_list == "Healthy Control"] <- "Healthy"
cohort_bl_list[cohort_bl_list == "Parkinson's Disease"] <- "PD"
# Read normalized data counts
sncrna_norm_df <- as.data.frame(fread("Data/PPMI_sncRNAcounts/counts/sncrna_quantification_matrix_rpm_norm.csv"))
rownames(sncrna_norm_df) <- sncrna_norm_df$Reference
sncrna_norm_df <- sncrna_norm_df[-1]
sncrna_norm_df_clean <- sncrna_norm_df[, c(PATNO_list %in% PATNO_list_clean)]
sncrna_norm_df_bl <- sncrna_norm_df_clean[, grepl("BL", colnames(sncrna_norm_df_clean))]
sncrna_norm_df_bl_t <- as.data.frame(t(sncrna_norm_df_bl))

bl_umap <- umap(sncrna_norm_df_bl_t, n_components = 2)
bl_umap_df <- data.frame("cohort" = cohort_bl_list, "UMAP1" = bl_umap$layout[,1], "UMAP2" = bl_umap$layout[,2])

bl_umap <- ggplot(
  bl_umap_df[!bl_umap_df$cohort == "SWEDD", ], 
  aes(x = UMAP1, y = UMAP2, color = cohort)
  ) + 
  geom_point(alpha = 0.9) + 
  labs(x = "UMAP 1", y = "UMAP 2") +
  ggtitle("UMAP: all sncRNAs at baseline") +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5)) +
  theme(
    plot.title = element_text(size = 15, face = "bold"),
    axis.title = element_text(size = 15),
    axis.text = element_text(size = 15),
    legend.title = element_text(size = 12), 
    legend.text = element_text(size = 12)
        ) +
  guides(color = guide_legend(override.aes = list(size = 4))) +
  theme(legend.title=element_blank())

bl_umap

# optionally save
#ggsave("figures/subplots/fig2_umap.png", 
#       plot = bl_umap,
#       width = 2000,
#       height = 1500,
#       units = "px")
```


```{r}
# LDA
# process data by removing SWEDD cohort
sncrna_df_bl_t_lda <- sncrna_df_bl_t[-which(cohort_bl_list == "SWEDD"), ]
# Run PCA
pca_results_for_lda <- prcomp(sncrna_norm_df_bl_t[-which(cohort_bl_list == "SWEDD"), ], center = TRUE)
# remove sncRNAs with zero values across all patients
sncrna_df_bl_t_pca <- sncrna_norm_df_bl_t[vapply(sncrna_norm_df_bl_t, function(x) length(unique(x)) > 1, logical(1L))]
pca_results_lda <- prcomp(sncrna_df_bl_t_pca[-which(cohort_bl_list == "SWEDD"), ], center = TRUE, scale = TRUE)
fviz_eig(pca_results_for_lda)
which(cohort_bl_list == "SWEDD")
pca_results_for_lda_top5 <- as.data.frame(pca_results_for_lda$x[ ,c(1:5)])
pca_results_for_lda_top5$cohort <- cohort_bl_list[-which(cohort_bl_list == "SWEDD")]
# Run LDA on top 5 PCs
lda_results <- lda(cohort ~., pca_results_for_lda_top5)
lda_predictions <- predict(lda_results)
lda_df <- data.frame("outcome" = pca_results_for_lda_top5$cohort,
                     "lda" = lda_predictions$x)
lda_plot <-
  ggplot(data = lda_df[!lda_df$lda.LD2 < -8, ], # remove single outlier 
       mapping = aes(x = lda.LD1, y = lda.LD2, color = outcome)) + 
  geom_point(size = 2, alpha = 0.7) +
  labs(x = "LD1", y = "LD2") +
  theme_classic() +
  stat_ellipse(aes(fill = outcome), geom = "polygon", alpha = 0.1) +
  geom_hline(yintercept = 0, size = .2) + 
  geom_vline(xintercept = 0, size = .2) +
  xlab("LD1") +
  ylab("LD2") +
  ggtitle("LDA: top 5 PCs from all sncRNAs at baseline") +
  theme(plot.title = element_text(hjust = 0.5)) +
    theme(
      plot.title = element_text(size = 15, face = "bold"),
      axis.title = element_text(size = 15),
      axis.text = element_text(size = 15),
      legend.title = element_text(size = 12), 
      legend.text = element_text(size = 12)
          ) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    theme(legend.title=element_blank())

lda_plot

# optionally save
#ggsave("figures/subplots/fig2_lda_plot.png", 
#       plot = lda_plot,
#       width = 2000,
#       height = 1500,
#       units = "px")
```


# Figure 2 - Differential expression analysis of sncRNAs between healthy, prodromal, and PD subjects at baseline

```{r}
# Prepare DE analysis
# Convert all columns to integers
sncrna_df_bl <- as.data.frame(lapply(sncrna_df_bl, as.integer))
rownames(sncrna_df_bl) <- sncrna_df$Reference

meta_df_bl <- data.frame("cohort" = cohort_bl_list)
rownames(meta_df_bl) <- colnames(sncrna_df_bl)

dds_bl <- DESeqDataSetFromMatrix(
  countData = sncrna_df_bl, 
  colData = meta_df_bl, 
  design = ~cohort
  )

dds_bl$cohort <- factor(dds_bl$cohort, levels = c("Healthy", "PD", "Prodromal", "SWEDD"))
dds_bl <- DESeq(dds_bl)
```


```{r}
# Parkinson's vs. healthy
res_bl_pd_h <- results(
  dds_bl, 
  contrast = c("cohort", "PD", "Healthy")
  )

res_bl_pd_h <- lfcShrink(
  dds_bl,
  contrast = c("cohort", "PD", "Healthy"), 
  res = res_bl_pd_h, type = "normal"
  )

res_bl_pd_h_df <- as.data.frame(res_bl_pd_h)

res_bl_pd_h_df <- read.csv("Data/res_bl_pd_h_df.csv")

volcano_bl_pd_h <- EnhancedVolcano(
   res_bl_pd_h_df,
   lab = sncrna_df$Reference,
   x = 'log2FoldChange',
   y = 'pvalue',
   title = 'PD vs. Healthy',
   subtitle = "sncRNASeq (at baseline)",
   caption = paste(nrow(res_bl_pd_h_df), "sncRNA transcripts analyzed"),
   pCutoff = 10e-5,
   FCcutoff = 1, # also tried 0.5
   pointSize = 3.0,
   labSize = 4.0,
   xlim = c(-2.5, 2.5)
 )

volcano_bl_pd_h

# optionally save
#ggsave("figures/subplots/fig3_volcano_bl_pd_h.png", 
#       plot = volcano_bl_pd_h,
#       width = 2500,
#       height = 3000,
#       units = "px")

# why are there many NA for padj = https://support.bioconductor.org/p/9135436/
```

```{r}
# PD vs. prodromal
res_bl_pd_p <- results(
  dds_bl, 
  contrast = c("cohort", "PD", "Prodromal")
  )

res_bl_pd_p <- lfcShrink(
  dds_bl,
  contrast = c("cohort", "PD", "Prodromal"), 
  res = res_bl_pd_p, type = "normal"
  )

summary(res_bl_pd_p)

res_bl_pd_p_df <- as.data.frame(res_bl_pd_p)

res_bl_pd_p_df <- read.csv("Data/res_bl_pd_p_df.csv")

volcano_bl_pd_p <- EnhancedVolcano(
   res_bl_pd_p_df,
   lab = sncrna_df$Reference,
   x = 'log2FoldChange',
   y = 'pvalue',
   title = 'PD vs. Prodromal',
   subtitle = "sncRNASeq (at baseline)",
   caption = paste(nrow(res_bl_pd_p_df), "sncRNA transcripts analyzed"),
   pCutoff = 10e-5,
   FCcutoff = 1, # also tried 0.5
   pointSize = 3.0,
   labSize = 4.0,
   xlim = c(-2.5, 2.5)
 )

volcano_bl_pd_p

# optionally save
#ggsave("figures/subplots/fig3_volcano_bl_pd_p.png", 
#       plot = volcano_bl_pd_p,
#       width = 2500,
#       height = 3000,
#       units = "px")
```

```{r}
# Prodromal vs. healthy
res_bl_p_h <- results(
  dds_bl, 
  contrast = c("cohort", "Prodromal", "Healthy")
  )

res_bl_p_h <- lfcShrink(
  dds_bl,
  contrast = c("cohort", "Prodromal", "Healthy"), 
  res = res_bl_p_h, type = "normal"
  )

summary(res_bl_p_h)

res_bl_p_h_df <- as.data.frame(res_bl_p_h)

res_bl_p_h_df <- read.csv("Data/res_bl_p_h_df.csv")

volcano_bl_p_h <- EnhancedVolcano(
   res_bl_p_h_df,
   lab = sncrna_df$Reference,
   x = 'log2FoldChange',
   y = 'pvalue',
   title = 'Prodromal vs. Healthy',
   subtitle = "sncRNASeq (at baseline)",
   caption = paste(nrow(res_bl_p_h_df), "sncRNA transcripts analyzed"),
   pCutoff = 10e-5,
   FCcutoff = 1, # also tried 0.5
   pointSize = 3.0,
   labSize = 4.0,
   xlim = c(-2.5, 2.5)
 )

volcano_bl_p_h

# optionally save
#ggsave("../figures/subplots/volcano_bl_p_h.png", 
#       plot = volcano_bl_p_h,
#       width = 2500,
#       height = 3000,
#      units = "px")
```


# Figure 3 - Number of differentially expressed sncRNAs in PD and prodromal subjects varies by age

```{r}
# Data processing to assign patients into age cohort by 10s of years
bl_patients <- gsub("\\..*","", substring(colnames(sncrna_norm_df_bl), 10))
bl_cohorts <- sncrna_cohort_full_list[grepl("BL", colnames(sncrna_norm_df_clean))]

bl_ages <- c()
for (i in 1:length(bl_patients)){
  bl_ages[i] <- participant_status_df$ENROLL_AGE[participant_status_df$PATNO == bl_patients[i]]
}

sncrna_df_bl_w_age <- sncrna_df_bl[, !is.na(bl_ages)]

bl_patients_w_age <- bl_patients[!is.na(bl_ages)]
bl_cohorts <- sncrna_cohort_full_list[grepl("BL", colnames(sncrna_norm_df_clean))]
bl_cohorts_w_age <- bl_cohorts[!is.na(bl_ages)]

bl_ages_clean <- round(bl_ages[!is.na(bl_ages)]) # round to nearest 10 age
bl_ages_group_clean <- round(bl_ages_clean/10)*10 # round to nearest 10 age

bl_demographics_df <- data.frame("patno" = bl_patients_w_age,
                                 "cohort" = bl_cohorts_w_age,
                                 "age" = bl_ages_clean,
                                 "age_group" = bl_ages_group_clean)

# Function to run DESeq2 on select cohort and age bracket
deseq_age_test <- function(cohort_of_interest, age_of_interest){
  exp_index <- which((bl_demographics_df$cohort == cohort_of_interest) & (bl_demographics_df$age == age_of_interest))
  control_index <- which((bl_demographics_df$cohort == "Healthy") & (bl_demographics_df$age == age_of_interest))
  exp_data <- sncrna_df_bl_w_age[,exp_index]
  control_data <- sncrna_df_bl_w_age[,control_index]
  both_data <- cbind(control_data, exp_data)
  meta_df <- data.frame("cohort" = c(rep("Healthy", ncol(control_data)), rep(cohort_of_interest, ncol(exp_data))))
  rownames(meta_df) <- c(colnames(control_data), colnames(exp_data))
  # Run DESeq2
  dds <- DESeqDataSetFromMatrix(
    countData = both_data, 
    colData = meta_df, 
    design = ~cohort
    )
  dds$cohort <- factor(dds$cohort, levels = c("Healthy", cohort_of_interest))
  dds_run <- DESeq(dds)
  results_dds <- results(
    dds_run, 
    contrast = c("cohort", cohort_of_interest, "Healthy")
    )
  results_dds_df <- as.data.frame(results_dds)
  return(results_dds_df) # return dataframe results of DESeq2
}

# PD DE analyses by age
pd_30_test <- deseq_age_test("PD", 30)
pd_40_test <- deseq_age_test("PD", 40)
pd_50_test <- deseq_age_test("PD", 50)
pd_60_test <- deseq_age_test("PD", 60)
pd_70_test <- deseq_age_test("PD", 70)
pd_80_test <- deseq_age_test("PD", 80)

# Prodromal DE analyses by age
p_40_test <- deseq_age_test("Prodromal", 40)
p_50_test <- deseq_age_test("Prodromal", 50)
p_60_test <- deseq_age_test("Prodromal", 60)
p_70_test <- deseq_age_test("Prodromal", 70)
p_80_test <- deseq_age_test("Prodromal", 80)

# Differentially expressed (up + down)
# Make dataframe of results
# P-value threshold: 10e-5
# log2 fold change cutoff: 1 +/- 0.25
dif_exp_age_df <- data.frame(
  "age" = c(40, 50, 60, 70, 80, 40, 50, 60, 70, 80),
  "cohort" = c(rep("PD", 5), rep("Prodromal", 5)),
  "num_transcripts_1" = c(
    length(pd_40_test$padj[!is.na(pd_40_test$padj)][(pd_40_test$padj[!is.na(pd_40_test$padj)] < 10e-5) & (abs(pd_40_test$log2FoldChange[!is.na(pd_40_test$padj)]) > 1)]),
    length(pd_50_test$padj[!is.na(pd_50_test$padj)][(pd_40_test$padj[!is.na(pd_50_test$padj)] < 10e-5) & (abs(pd_50_test$log2FoldChange[!is.na(pd_50_test$padj)]) > 1)]),
    length(pd_60_test$padj[!is.na(pd_60_test$padj)][(pd_40_test$padj[!is.na(pd_60_test$padj)] < 10e-5) & (abs(pd_60_test$log2FoldChange[!is.na(pd_60_test$padj)]) > 1)]),
    length(pd_70_test$padj[!is.na(pd_70_test$padj)][(pd_40_test$padj[!is.na(pd_70_test$padj)] < 10e-5) & (abs(pd_70_test$log2FoldChange[!is.na(pd_70_test$padj)]) > 1)]),
    length(pd_80_test$padj[!is.na(pd_80_test$padj)][(pd_40_test$padj[!is.na(pd_80_test$padj)] < 10e-5) & (abs(pd_80_test$log2FoldChange[!is.na(pd_80_test$padj)]) > 1)]),
    length(p_40_test$padj[!is.na(p_40_test$padj)][(p_40_test$padj[!is.na(p_40_test$padj)] < 10e-5) & (abs(p_40_test$log2FoldChange[!is.na(p_40_test$padj)]) > 1)]),
    length(p_50_test$padj[!is.na(p_50_test$padj)][(p_40_test$padj[!is.na(p_50_test$padj)] < 10e-5) & (abs(p_50_test$log2FoldChange[!is.na(p_50_test$padj)]) > 1)]),
    length(p_60_test$padj[!is.na(p_60_test$padj)][(p_40_test$padj[!is.na(p_60_test$padj)] < 10e-5) & (abs(p_60_test$log2FoldChange[!is.na(p_60_test$padj)]) > 1)]),
    length(p_70_test$padj[!is.na(p_70_test$padj)][(p_40_test$padj[!is.na(p_70_test$padj)] < 10e-5) & (abs(p_70_test$log2FoldChange[!is.na(p_70_test$padj)]) > 1)]),
    length(p_80_test$padj[!is.na(p_80_test$padj)][(p_40_test$padj[!is.na(p_80_test$padj)] < 10e-5) & (abs(p_80_test$log2FoldChange[!is.na(p_80_test$padj)]) > 1)])
  ),
  "num_transcripts_125" = c(
    length(pd_40_test$padj[!is.na(pd_40_test$padj)][(pd_40_test$padj[!is.na(pd_40_test$padj)] < 10e-5) & (abs(pd_40_test$log2FoldChange[!is.na(pd_40_test$padj)]) > 1.25)]),
    length(pd_50_test$padj[!is.na(pd_50_test$padj)][(pd_40_test$padj[!is.na(pd_50_test$padj)] < 10e-5) & (abs(pd_50_test$log2FoldChange[!is.na(pd_50_test$padj)]) > 1.25)]),
    length(pd_60_test$padj[!is.na(pd_60_test$padj)][(pd_40_test$padj[!is.na(pd_60_test$padj)] < 10e-5) & (abs(pd_60_test$log2FoldChange[!is.na(pd_60_test$padj)]) > 1.25)]),
    length(pd_70_test$padj[!is.na(pd_70_test$padj)][(pd_40_test$padj[!is.na(pd_70_test$padj)] < 10e-5) & (abs(pd_70_test$log2FoldChange[!is.na(pd_70_test$padj)]) > 1.25)]),
    length(pd_80_test$padj[!is.na(pd_80_test$padj)][(pd_40_test$padj[!is.na(pd_80_test$padj)] < 10e-5) & (abs(pd_80_test$log2FoldChange[!is.na(pd_80_test$padj)]) > 1.25)]),
    length(p_40_test$padj[!is.na(p_40_test$padj)][(p_40_test$padj[!is.na(p_40_test$padj)] < 10e-5) & (abs(p_40_test$log2FoldChange[!is.na(p_40_test$padj)]) > 1.25)]),
    length(p_50_test$padj[!is.na(p_50_test$padj)][(p_40_test$padj[!is.na(p_50_test$padj)] < 10e-5) & (abs(p_50_test$log2FoldChange[!is.na(p_50_test$padj)]) > 1.25)]),
    length(p_60_test$padj[!is.na(p_60_test$padj)][(p_40_test$padj[!is.na(p_60_test$padj)] < 10e-5) & (abs(p_60_test$log2FoldChange[!is.na(p_60_test$padj)]) > 1.25)]),
    length(p_70_test$padj[!is.na(p_70_test$padj)][(p_40_test$padj[!is.na(p_70_test$padj)] < 10e-5) & (abs(p_70_test$log2FoldChange[!is.na(p_70_test$padj)]) > 1.25)]),
    length(p_80_test$padj[!is.na(p_80_test$padj)][(p_40_test$padj[!is.na(p_80_test$padj)] < 10e-5) & (abs(p_80_test$log2FoldChange[!is.na(p_80_test$padj)]) > 1.25)])
  ),
  "num_transcripts_075" = c(
    length(pd_40_test$padj[!is.na(pd_40_test$padj)][(pd_40_test$padj[!is.na(pd_40_test$padj)] < 10e-5) & (abs(pd_40_test$log2FoldChange[!is.na(pd_40_test$padj)]) > 0.75)]),
    length(pd_50_test$padj[!is.na(pd_50_test$padj)][(pd_40_test$padj[!is.na(pd_50_test$padj)] < 10e-5) & (abs(pd_50_test$log2FoldChange[!is.na(pd_50_test$padj)]) > 0.75)]),
    length(pd_60_test$padj[!is.na(pd_60_test$padj)][(pd_40_test$padj[!is.na(pd_60_test$padj)] < 10e-5) & (abs(pd_60_test$log2FoldChange[!is.na(pd_60_test$padj)]) > 0.75)]),
    length(pd_70_test$padj[!is.na(pd_70_test$padj)][(pd_40_test$padj[!is.na(pd_70_test$padj)] < 10e-5) & (abs(pd_70_test$log2FoldChange[!is.na(pd_70_test$padj)]) > 0.75)]),
    length(pd_80_test$padj[!is.na(pd_80_test$padj)][(pd_40_test$padj[!is.na(pd_80_test$padj)] < 10e-5) & (abs(pd_80_test$log2FoldChange[!is.na(pd_80_test$padj)]) > 0.75)]),
    length(p_40_test$padj[!is.na(p_40_test$padj)][(p_40_test$padj[!is.na(p_40_test$padj)] < 10e-5) & (abs(p_40_test$log2FoldChange[!is.na(p_40_test$padj)]) > 0.75)]),
    length(p_50_test$padj[!is.na(p_50_test$padj)][(p_40_test$padj[!is.na(p_50_test$padj)] < 10e-5) & (abs(p_50_test$log2FoldChange[!is.na(p_50_test$padj)]) > 0.75)]),
    length(p_60_test$padj[!is.na(p_60_test$padj)][(p_40_test$padj[!is.na(p_60_test$padj)] < 10e-5) & (abs(p_60_test$log2FoldChange[!is.na(p_60_test$padj)]) > 0.75)]),
    length(p_70_test$padj[!is.na(p_70_test$padj)][(p_40_test$padj[!is.na(p_70_test$padj)] < 10e-5) & (abs(p_70_test$log2FoldChange[!is.na(p_70_test$padj)]) > 0.75)]),
    length(p_80_test$padj[!is.na(p_80_test$padj)][(p_40_test$padj[!is.na(p_80_test$padj)] < 10e-5) & (abs(p_80_test$log2FoldChange[!is.na(p_80_test$padj)]) > 0.75)])
  )
)

# Overexpressed
# Make dataframe of results
# P-value threshold: 10e-5
# log2 fold change cutoff: 1 +/- 0.25
up_exp_age_df <- data.frame(
  "age" = c(40, 50, 60, 70, 80, 40, 50, 60, 70, 80),
  "cohort" = c(rep("PD", 5), rep("Prodromal", 5)),
  "num_transcripts_1" = c(
    length(pd_40_test$padj[!is.na(pd_40_test$padj)][(pd_40_test$padj[!is.na(pd_40_test$padj)] < 10e-5) & (pd_40_test$log2FoldChange[!is.na(pd_40_test$padj)] > 1)]),
    length(pd_50_test$padj[!is.na(pd_50_test$padj)][(pd_40_test$padj[!is.na(pd_50_test$padj)] < 10e-5) & (pd_50_test$log2FoldChange[!is.na(pd_50_test$padj)] > 1)]),
    length(pd_60_test$padj[!is.na(pd_60_test$padj)][(pd_40_test$padj[!is.na(pd_60_test$padj)] < 10e-5) & (pd_60_test$log2FoldChange[!is.na(pd_60_test$padj)] > 1)]),
    length(pd_70_test$padj[!is.na(pd_70_test$padj)][(pd_40_test$padj[!is.na(pd_70_test$padj)] < 10e-5) & (pd_70_test$log2FoldChange[!is.na(pd_70_test$padj)] > 1)]),
    length(pd_80_test$padj[!is.na(pd_80_test$padj)][(pd_40_test$padj[!is.na(pd_80_test$padj)] < 10e-5) & (pd_80_test$log2FoldChange[!is.na(pd_80_test$padj)] > 1)]),
    length(p_40_test$padj[!is.na(p_40_test$padj)][(p_40_test$padj[!is.na(p_40_test$padj)] < 10e-5) & (p_40_test$log2FoldChange[!is.na(p_40_test$padj)] > 1)]),
    length(p_50_test$padj[!is.na(p_50_test$padj)][(p_40_test$padj[!is.na(p_50_test$padj)] < 10e-5) & (p_50_test$log2FoldChange[!is.na(p_50_test$padj)] > 1)]),
    length(p_60_test$padj[!is.na(p_60_test$padj)][(p_40_test$padj[!is.na(p_60_test$padj)] < 10e-5) & (p_60_test$log2FoldChange[!is.na(p_60_test$padj)] > 1)]),
    length(p_70_test$padj[!is.na(p_70_test$padj)][(p_40_test$padj[!is.na(p_70_test$padj)] < 10e-5) & (p_70_test$log2FoldChange[!is.na(p_70_test$padj)] > 1)]),
    length(p_80_test$padj[!is.na(p_80_test$padj)][(p_40_test$padj[!is.na(p_80_test$padj)] < 10e-5) & (p_80_test$log2FoldChange[!is.na(p_80_test$padj)] > 1)])
  ),
  "num_transcripts_125" = c(
    length(pd_40_test$padj[!is.na(pd_40_test$padj)][(pd_40_test$padj[!is.na(pd_40_test$padj)] < 10e-5) & (pd_40_test$log2FoldChange[!is.na(pd_40_test$padj)] > 1.25)]),
    length(pd_50_test$padj[!is.na(pd_50_test$padj)][(pd_40_test$padj[!is.na(pd_50_test$padj)] < 10e-5) & (pd_50_test$log2FoldChange[!is.na(pd_50_test$padj)] > 1.25)]),
    length(pd_60_test$padj[!is.na(pd_60_test$padj)][(pd_40_test$padj[!is.na(pd_60_test$padj)] < 10e-5) & (pd_60_test$log2FoldChange[!is.na(pd_60_test$padj)] > 1.25)]),
    length(pd_70_test$padj[!is.na(pd_70_test$padj)][(pd_40_test$padj[!is.na(pd_70_test$padj)] < 10e-5) & (pd_70_test$log2FoldChange[!is.na(pd_70_test$padj)] > 1.25)]),
    length(pd_80_test$padj[!is.na(pd_80_test$padj)][(pd_40_test$padj[!is.na(pd_80_test$padj)] < 10e-5) & (pd_80_test$log2FoldChange[!is.na(pd_80_test$padj)] > 1.25)]),
    length(p_40_test$padj[!is.na(p_40_test$padj)][(p_40_test$padj[!is.na(p_40_test$padj)] < 10e-5) & (p_40_test$log2FoldChange[!is.na(p_40_test$padj)] > 1.25)]),
    length(p_50_test$padj[!is.na(p_50_test$padj)][(p_40_test$padj[!is.na(p_50_test$padj)] < 10e-5) & (p_50_test$log2FoldChange[!is.na(p_50_test$padj)] > 1.25)]),
    length(p_60_test$padj[!is.na(p_60_test$padj)][(p_40_test$padj[!is.na(p_60_test$padj)] < 10e-5) & (p_60_test$log2FoldChange[!is.na(p_60_test$padj)] > 1.25)]),
    length(p_70_test$padj[!is.na(p_70_test$padj)][(p_40_test$padj[!is.na(p_70_test$padj)] < 10e-5) & (p_70_test$log2FoldChange[!is.na(p_70_test$padj)] > 1.25)]),
    length(p_80_test$padj[!is.na(p_80_test$padj)][(p_40_test$padj[!is.na(p_80_test$padj)] < 10e-5) & (p_80_test$log2FoldChange[!is.na(p_80_test$padj)] > 1.25)])
  ),
  "num_transcripts_075" = c(
    length(pd_40_test$padj[!is.na(pd_40_test$padj)][(pd_40_test$padj[!is.na(pd_40_test$padj)] < 10e-5) & (pd_40_test$log2FoldChange[!is.na(pd_40_test$padj)] > 0.75)]),
    length(pd_50_test$padj[!is.na(pd_50_test$padj)][(pd_40_test$padj[!is.na(pd_50_test$padj)] < 10e-5) & (pd_50_test$log2FoldChange[!is.na(pd_50_test$padj)] > 0.75)]),
    length(pd_60_test$padj[!is.na(pd_60_test$padj)][(pd_40_test$padj[!is.na(pd_60_test$padj)] < 10e-5) & (pd_60_test$log2FoldChange[!is.na(pd_60_test$padj)] > 0.75)]),
    length(pd_70_test$padj[!is.na(pd_70_test$padj)][(pd_40_test$padj[!is.na(pd_70_test$padj)] < 10e-5) & (pd_70_test$log2FoldChange[!is.na(pd_70_test$padj)] > 0.75)]),
    length(pd_80_test$padj[!is.na(pd_80_test$padj)][(pd_40_test$padj[!is.na(pd_80_test$padj)] < 10e-5) & (pd_80_test$log2FoldChange[!is.na(pd_80_test$padj)] > 0.75)]),
    length(p_40_test$padj[!is.na(p_40_test$padj)][(p_40_test$padj[!is.na(p_40_test$padj)] < 10e-5) & (p_40_test$log2FoldChange[!is.na(p_40_test$padj)] > 0.75)]),
    length(p_50_test$padj[!is.na(p_50_test$padj)][(p_40_test$padj[!is.na(p_50_test$padj)] < 10e-5) & (p_50_test$log2FoldChange[!is.na(p_50_test$padj)] > 0.75)]),
    length(p_60_test$padj[!is.na(p_60_test$padj)][(p_40_test$padj[!is.na(p_60_test$padj)] < 10e-5) & (p_60_test$log2FoldChange[!is.na(p_60_test$padj)] > 0.75)]),
    length(p_70_test$padj[!is.na(p_70_test$padj)][(p_40_test$padj[!is.na(p_70_test$padj)] < 10e-5) & (p_70_test$log2FoldChange[!is.na(p_70_test$padj)] > 0.75)]),
    length(p_80_test$padj[!is.na(p_80_test$padj)][(p_40_test$padj[!is.na(p_80_test$padj)] < 10e-5) & (p_80_test$log2FoldChange[!is.na(p_80_test$padj)] > 0.75)])
  )
)

# Underexpressed
# Make dataframe of results
# P-value threshold: 10e-5
# log2 fold change cutoff: 1 +/- 0.25
down_exp_age_df <- data.frame(
  "age" = c(40, 50, 60, 70, 80, 40, 50, 60, 70, 80),
  "cohort" = c(rep("PD", 5), rep("Prodromal", 5)),
  "num_transcripts_1" = c(
    length(pd_40_test$padj[!is.na(pd_40_test$padj)][(pd_40_test$padj[!is.na(pd_40_test$padj)] < 10e-5) & (pd_40_test$log2FoldChange[!is.na(pd_40_test$padj)] < -1)]),
    length(pd_50_test$padj[!is.na(pd_50_test$padj)][(pd_40_test$padj[!is.na(pd_50_test$padj)] < 10e-5) & (pd_50_test$log2FoldChange[!is.na(pd_50_test$padj)] < -1)]),
    length(pd_60_test$padj[!is.na(pd_60_test$padj)][(pd_40_test$padj[!is.na(pd_60_test$padj)] < 10e-5) & (pd_60_test$log2FoldChange[!is.na(pd_60_test$padj)] < -1)]),
    length(pd_70_test$padj[!is.na(pd_70_test$padj)][(pd_40_test$padj[!is.na(pd_70_test$padj)] < 10e-5) & (pd_70_test$log2FoldChange[!is.na(pd_70_test$padj)] < -1)]),
    length(pd_80_test$padj[!is.na(pd_80_test$padj)][(pd_40_test$padj[!is.na(pd_80_test$padj)] < 10e-5) & (pd_80_test$log2FoldChange[!is.na(pd_80_test$padj)] < -1)]),
    length(p_40_test$padj[!is.na(p_40_test$padj)][(p_40_test$padj[!is.na(p_40_test$padj)] < 10e-5) & (p_40_test$log2FoldChange[!is.na(p_40_test$padj)] < -1)]),
    length(p_50_test$padj[!is.na(p_50_test$padj)][(p_40_test$padj[!is.na(p_50_test$padj)] < 10e-5) & (p_50_test$log2FoldChange[!is.na(p_50_test$padj)] < -1)]),
    length(p_60_test$padj[!is.na(p_60_test$padj)][(p_40_test$padj[!is.na(p_60_test$padj)] < 10e-5) & (p_60_test$log2FoldChange[!is.na(p_60_test$padj)] < -1)]),
    length(p_70_test$padj[!is.na(p_70_test$padj)][(p_40_test$padj[!is.na(p_70_test$padj)] < 10e-5) & (p_70_test$log2FoldChange[!is.na(p_70_test$padj)] < -1)]),
    length(p_80_test$padj[!is.na(p_80_test$padj)][(p_40_test$padj[!is.na(p_80_test$padj)] < 10e-5) & (p_80_test$log2FoldChange[!is.na(p_80_test$padj)] < -1)])
  ),
  "num_transcripts_125" = c(
    length(pd_40_test$padj[!is.na(pd_40_test$padj)][(pd_40_test$padj[!is.na(pd_40_test$padj)] < 10e-5) & (pd_40_test$log2FoldChange[!is.na(pd_40_test$padj)] < -1.25)]),
    length(pd_50_test$padj[!is.na(pd_50_test$padj)][(pd_40_test$padj[!is.na(pd_50_test$padj)] < 10e-5) & (pd_50_test$log2FoldChange[!is.na(pd_50_test$padj)] < -1.25)]),
    length(pd_60_test$padj[!is.na(pd_60_test$padj)][(pd_40_test$padj[!is.na(pd_60_test$padj)] < 10e-5) & (pd_60_test$log2FoldChange[!is.na(pd_60_test$padj)] < -1.25)]),
    length(pd_70_test$padj[!is.na(pd_70_test$padj)][(pd_40_test$padj[!is.na(pd_70_test$padj)] < 10e-5) & (pd_70_test$log2FoldChange[!is.na(pd_70_test$padj)] < -1.25)]),
    length(pd_80_test$padj[!is.na(pd_80_test$padj)][(pd_40_test$padj[!is.na(pd_80_test$padj)] < 10e-5) & (pd_80_test$log2FoldChange[!is.na(pd_80_test$padj)] < -1.25)]),
    length(p_40_test$padj[!is.na(p_40_test$padj)][(p_40_test$padj[!is.na(p_40_test$padj)] < 10e-5) & (p_40_test$log2FoldChange[!is.na(p_40_test$padj)] < -1.25)]),
    length(p_50_test$padj[!is.na(p_50_test$padj)][(p_40_test$padj[!is.na(p_50_test$padj)] < 10e-5) & (p_50_test$log2FoldChange[!is.na(p_50_test$padj)] < -1.25)]),
    length(p_60_test$padj[!is.na(p_60_test$padj)][(p_40_test$padj[!is.na(p_60_test$padj)] < 10e-5) & (p_60_test$log2FoldChange[!is.na(p_60_test$padj)] < -1.25)]),
    length(p_70_test$padj[!is.na(p_70_test$padj)][(p_40_test$padj[!is.na(p_70_test$padj)] < 10e-5) & (p_70_test$log2FoldChange[!is.na(p_70_test$padj)] < -1.25)]),
    length(p_80_test$padj[!is.na(p_80_test$padj)][(p_40_test$padj[!is.na(p_80_test$padj)] < 10e-5) & (p_80_test$log2FoldChange[!is.na(p_80_test$padj)] < -1.25)])
  ),
  "num_transcripts_075" = c(
    length(pd_40_test$padj[!is.na(pd_40_test$padj)][(pd_40_test$padj[!is.na(pd_40_test$padj)] < 10e-5) & (pd_40_test$log2FoldChange[!is.na(pd_40_test$padj)] < -0.75)]),
    length(pd_50_test$padj[!is.na(pd_50_test$padj)][(pd_40_test$padj[!is.na(pd_50_test$padj)] < 10e-5) & (pd_50_test$log2FoldChange[!is.na(pd_50_test$padj)] < -0.75)]),
    length(pd_60_test$padj[!is.na(pd_60_test$padj)][(pd_40_test$padj[!is.na(pd_60_test$padj)] < 10e-5) & (pd_60_test$log2FoldChange[!is.na(pd_60_test$padj)] < -0.75)]),
    length(pd_70_test$padj[!is.na(pd_70_test$padj)][(pd_40_test$padj[!is.na(pd_70_test$padj)] < 10e-5) & (pd_70_test$log2FoldChange[!is.na(pd_70_test$padj)] < -0.75)]),
    length(pd_80_test$padj[!is.na(pd_80_test$padj)][(pd_40_test$padj[!is.na(pd_80_test$padj)] < 10e-5) & (pd_80_test$log2FoldChange[!is.na(pd_80_test$padj)] < -0.75)]),
    length(p_40_test$padj[!is.na(p_40_test$padj)][(p_40_test$padj[!is.na(p_40_test$padj)] < 10e-5) & (p_40_test$log2FoldChange[!is.na(p_40_test$padj)] < -0.75)]),
    length(p_50_test$padj[!is.na(p_50_test$padj)][(p_40_test$padj[!is.na(p_50_test$padj)] < 10e-5) & (p_50_test$log2FoldChange[!is.na(p_50_test$padj)] < -0.75)]),
    length(p_60_test$padj[!is.na(p_60_test$padj)][(p_40_test$padj[!is.na(p_60_test$padj)] < 10e-5) & (p_60_test$log2FoldChange[!is.na(p_60_test$padj)] < -0.75)]),
    length(p_70_test$padj[!is.na(p_70_test$padj)][(p_40_test$padj[!is.na(p_70_test$padj)] < 10e-5) & (p_70_test$log2FoldChange[!is.na(p_70_test$padj)] < -0.75)]),
    length(p_80_test$padj[!is.na(p_80_test$padj)][(p_40_test$padj[!is.na(p_80_test$padj)] < 10e-5) & (p_80_test$log2FoldChange[!is.na(p_80_test$padj)] < -0.75)])
  )
)

# Plots
diff_plot <-
  ggplot(dif_exp_age_df,
       aes(x = age,
           y = num_transcripts_1,
           ymin = num_transcripts_075,
           ymax = num_transcripts_125,
           fill = cohort,
           color = cohort)) +
  geom_line(linewidth = 1.5) +
  geom_ribbon(alpha = 0.3) +
  geom_point(aes(color = cohort), size = 3) + 
  ggtitle("# Diff. Expressed sncRNAs (compared to Healthy)") +
  xlab("Age (years)") +
  ylab("Count") +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5)) +
  labs(color = NULL) +
  labs(fill = NULL)

up_plot <- 
  ggplot(up_exp_age_df,
       aes(x = age,
           y = num_transcripts_1,
           ymin = num_transcripts_075,
           ymax = num_transcripts_125,
           fill = cohort,
           color = cohort)) +
  geom_line(linewidth = 1.5) +
  geom_ribbon(alpha = 0.3) +
  geom_point(aes(color = cohort), size = 3) + 
  ggtitle("# Overexpressed sncRNAs") +
  xlab("Age (years)") +
  ylab("Count") +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5)) +
  labs(color = NULL) +
  labs(fill = NULL)

down_plot <- 
  ggplot(down_exp_age_df,
       aes(x = age,
           y = num_transcripts_1,
           ymin = num_transcripts_075,
           ymax = num_transcripts_125,
           fill = cohort,
           color = cohort)) +
  geom_line(linewidth = 1.5) +
  geom_ribbon(alpha = 0.3) +
  geom_point(aes(color = cohort), size = 3) + 
  ggtitle("# Underexpressed sncRNAs") +
  xlab("Age (years)") +
  ylab("Count") +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5)) +
  labs(color = NULL) +
  labs(fill = NULL)

# Combine DE, overexpressed, and underexpressed in one plot
age_plot <- ggarrange(diff_plot, up_plot, down_plot, ncol = 3, common.legend = TRUE, legend = "bottom")

age_plot

# optionally save
#ggsave("figures/subplots/fig3_age_plot.png", 
#       plot = age_plot,
#       width = 4500,
#       height = 1500,
#       units = "px")
```


# Figure 4 - SNORD48 is amplified in PD patients with worsening motor symptoms over time  

```{r}
# Line plot of motor scores over time for PD subjects
# Load UPDRS Part III clinical score data
updrs_part3_df <- read.csv("Data/MDS-UPDRS_Part_III.csv")
# Only consider PD patients
sncrna_norm_df_pd <- sncrna_norm_df_clean[,which(sncrna_cohort_full_list == "PD")]
pd_patno <- unique(gsub("\\..*", "", substring(colnames(sncrna_norm_df_pd), 10)))
# Only consider PD patients with at least 5 visits
num_visits_p <- c()
for (i in 1:length(pd_patno)){
  num_visits_p[i] <- table(grepl(paste0("\\.",pd_patno[i], "\\."), colnames(sncrna_norm_df_pd)))["TRUE"]
}
pd_patno_5plus_visits <- pd_patno[which(num_visits_p >= 5)] # relax to 5 in attempt 2
sncrna_pd_5plus_df <- sncrna_df_clean[, grepl(paste(pd_patno_5plus_visits, collapse = "|"), colnames(sncrna_df_clean))]
updrs_part3_pd_patno <- unique(updrs_part3_df$PATNO[which(updrs_part3_df$PATNO %in% pd_patno_5plus_visits)])
updrs_part3_pd <- list()
for (i in 1:length(updrs_part3_pd_patno)){
  updrs_part3_pd[[i]] <- updrs_part3_df$NP3TOT[updrs_part3_df$PATNO == updrs_part3_pd_patno[i]]
}
# Get visit index
visits_index <- list()
updrs_part3_pd_with_tx_data <- list()
for (i in 1:length(updrs_part3_pd_patno)){ # for each patient
  colnames_i <- colnames(sncrna_norm_df_pd)[grepl(paste0("\\.",updrs_part3_pd_patno[i], "\\."), colnames(sncrna_norm_df_pd))]
  visits_index[[i]]  <- substr(sub(".*.V", "", colnames_i), start = 1, stop = 2)
  for (j in 1:length(visits_index[[i]])){
    if (visits_index[[i]][j] == "PP"){
      visits_index[[i]][j] <- as.numeric(1)
    } else {
      visits_index[[i]][j] <- as.numeric(visits_index[[i]][j])
    }
  }
  visits_index[[i]] <- as.numeric(visits_index[[i]])
  updrs_part3_pd_with_tx_data[[i]] <- nafill(updrs_part3_pd[[i]][visits_index[[i]]], type = "locf")
  updrs_part3_pd_with_tx_data[[i]] <- nafill(updrs_part3_pd_with_tx_data[[i]], type = "nocb")
}

# Have a +/- 8 change in UPDRS Part III score over the 5 visits
# to be considered improved or worsened, respectively
updrs_part3_pd_with_tx_data_change_large <- c()
for (i in 1:length(updrs_part3_pd_with_tx_data)){
  if (updrs_part3_pd_with_tx_data[[i]][length(updrs_part3_pd_with_tx_data[[i]])] - updrs_part3_pd_with_tx_data[[i]][1] > 8){
    updrs_part3_pd_with_tx_data_change_large[i] <- "worsened"
  } else if (updrs_part3_pd_with_tx_data[[i]][length(updrs_part3_pd_with_tx_data[[i]])] - updrs_part3_pd_with_tx_data[[i]][1] < -8){
    updrs_part3_pd_with_tx_data_change_large[i] <- "improved"
  } else{
    updrs_part3_pd_with_tx_data_change_large[i] <- "no_change"
  }
}

updrs_part3_pd_with_tx_data_df <- as.data.frame(updrs_part3_pd_with_tx_data)
colnames(updrs_part3_pd_with_tx_data_df) <- c(1:ncol(updrs_part3_pd_with_tx_data_df))
updrs_part3_pd_with_tx_data_df$visit <- c("V1 (BL)", "V2", "V3", "V4", "V5")
updrs_part3_pd_with_tx_data_df_v2 <- melt(updrs_part3_pd_with_tx_data_df,  id.vars = "visit", variable.name = "series")
updrs_part3_pd_with_tx_data_df_v2$`Motor symptoms` <- NA
for (i in 1:nrow(updrs_part3_pd_with_tx_data_df_v2)){
  if (updrs_part3_pd_with_tx_data_df_v2$series[i] %in% which(updrs_part3_pd_with_tx_data_change_large == "improved")){
    updrs_part3_pd_with_tx_data_df_v2$`Motor symptoms`[i] <- "Improved"
  } else if (updrs_part3_pd_with_tx_data_df_v2$series[i] %in% which(updrs_part3_pd_with_tx_data_change_large == "worsened")){
    updrs_part3_pd_with_tx_data_df_v2$`Motor symptoms`[i] <- "Worsened"
  } else if (updrs_part3_pd_with_tx_data_df_v2$series[i] %in% which(updrs_part3_pd_with_tx_data_change_large == "no_change")){
    updrs_part3_pd_with_tx_data_df_v2$`Motor symptoms`[i] <- "No change"
  }
}

updrs_imp_wor_line_plot <-
ggplot(updrs_part3_pd_with_tx_data_df_v2, aes(x = visit, value)) + 
  geom_line(aes(group = series, color = `Motor symptoms`), alpha = 0.5, size = 0.7) +
  scale_color_manual(values=c("#00bf29", "#c2c2c2", "#ff0000")) +
  ggtitle("Motor score of PD subjects over 5 visits") +
  ylab("Motor impairment score (UPDRS III)") +
  xlab("Visit") + 
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5)) +
    theme(
      plot.title = element_text(size = 15, face = "bold"),
      axis.title = element_text(size = 15),
      axis.text = element_text(size = 15),
      legend.title = element_text(size = 12), 
      legend.text = element_text(size = 12)
          ) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    theme(legend.title=element_blank())

updrs_imp_wor_line_plot

# optionally save
#ggsave("figures/subplots/fig4_updrs_imp_wor_line_plot.png", 
#       plot = updrs_imp_wor_line_plot,
#       width = 2000,
#       height = 1300,
#       units = "px")
```

```{r}
# Differential expression volcano plot for... 
# PD subjects w/ worsening vs. unchanging motor symptoms
# Get indices for improved, worsened, and no change PD subjects
patno_norm_df_pd <- gsub("\\..*","", substring(colnames(sncrna_norm_df_pd)[-1], 10))
imp_indices <- which(patno_norm_df_pd %in% updrs_part3_pd_patno[which(updrs_part3_pd_with_tx_data_change_large == "improved")])
wor_indices <- which(patno_norm_df_pd %in% updrs_part3_pd_patno[which(updrs_part3_pd_with_tx_data_change_large == "worsened")])
nc_indices <- which(patno_norm_df_pd %in% updrs_part3_pd_patno[which(updrs_part3_pd_with_tx_data_change_large == "no_change")])

sncrna_norm_df_pd_imp <- sncrna_norm_df_pd[,imp_indices]
sncrna_norm_df_pd_imp_BL <- sncrna_norm_df_pd_imp[, grepl("BL", colnames(sncrna_norm_df_pd_imp))]

sncrna_norm_df_pd_wor <- sncrna_norm_df_pd[,wor_indices]
sncrna_norm_df_pd_wor_BL <- sncrna_norm_df_pd_wor[, grepl("BL", colnames(sncrna_norm_df_pd_wor))]

sncrna_norm_df_pd_nc <- sncrna_norm_df_pd[,nc_indices]
sncrna_norm_df_pd_nc_BL <- sncrna_norm_df_pd_nc[, grepl("BL", colnames(sncrna_norm_df_pd_nc))]

# Combine improved, worsened, and no change PD subjects into one dataframe
sncrna_norm_df_pd_imp_wor_nc_BL <- cbind(sncrna_norm_df_pd_imp_BL, sncrna_norm_df_pd_wor_BL, sncrna_norm_df_pd_nc_BL)
sncrna_norm_df_pd_imp_wor_nc_BL <- as.data.frame(lapply(sncrna_norm_df_pd_imp_wor_nc_BL, as.integer))

meta_df <- data.frame("cohort" = c(rep("Improved", ncol(sncrna_norm_df_pd_imp_BL)), rep("Worsened", ncol(sncrna_norm_df_pd_wor_BL)), rep("No_change", ncol(sncrna_norm_df_pd_nc_BL))))
rownames(meta_df) <- colnames(sncrna_norm_df_pd_imp_wor_nc_BL)

dds <- DESeqDataSetFromMatrix(
  countData = sncrna_norm_df_pd_imp_wor_nc_BL, 
  colData = meta_df, 
  design = ~cohort
  )
dds$cohort <- factor(dds$cohort, levels = c("Improved", "Worsened", "No_change"))

dds_run <- DESeq(dds)

results_dds_wor_nc <- results(
  dds_run, 
  contrast = c("cohort", "Worsened", "No_change")
  )

#results_dds_wor_im <- results(
#  dds_run, 
#  contrast = c("cohort", "Worsened", "Improved")
#  )

results_dds_wor_nc_df <- as.data.frame(results_dds_wor_nc)
#results_dds_wor_im_df <- as.data.frame(results_dds_wor_im)

volcano_wor_nc <- 
  EnhancedVolcano(
   results_dds_wor_nc_df,
   lab = sncrna_df$Reference,
   x = 'log2FoldChange',
   y = 'pvalue',
   title = 'PD subjects w/ worsening vs. unchanging motor symptoms',
   subtitle = "sncRNASeq (at baseline)",
   caption = paste(nrow(sncrna_df), "sncRNA transcripts analyzed"),
   pCutoff = 10e-5,
   FCcutoff = 1, # also tried 0.5
   pointSize = 3.0,
   labSize = 4.0,
   xlim = c(-2.5, 2.5)
 )

volcano_wor_nc

# optionally save
#ggsave("figures/subplots/fig4_volcano_wor_nc.png", 
#       plot = volcano_wor_nc,
#       width = 2500,
#       height = 3000,
#       units = "px")
```

# Supplementary 1 - PCA biplot from clinical data - PD vs. prodromal subjects

```{r}
# Figure 1 - PCA biplot from clinical data for PD and prodromal subjects

# Load .csv datasets for prodromal cohort and Parkinson's disease (PD) cohort
prodromal_df <- read.csv("Data/PPMI_Prodromal_Cohort_BL_to_Year_1_Dataset_Apr2020.csv")
pd_df <- read.csv("Data/PPMI_Original_Cohort_BL_to_Year_5_Dataset_Apr2020.csv")

# Keep columns that are shared between the two datasets
pd_df_shared <- pd_df[, colnames(pd_df) %in% colnames(prodromal_df)]
prodromal_df_shared <- prodromal_df[, colnames(prodromal_df) %in% colnames(pd_df)]

# Add explicit cohort column
pd_df_shared$cohort <- "PD" # has parkinson's disease
prodromal_df_shared$cohort <- "Prodromal" # prodromal, i.e., at risk for PD

# Combine the two dataframes
pd_prodromal_df <- rbind(pd_df_shared, prodromal_df_shared)

# Keep only baseline measurement values for each patient
pd_prodromal_df <- pd_prodromal_df[pd_prodromal_df$EVENT_ID == "BL", ]

# Remove some categorical variables not important for factor analysis
pd_prodromal_df <- pd_prodromal_df[, !colnames(pd_prodromal_df) %in% c("SITE", 
                                                                       "APPRDX", 
                                                                       "EVENT_ID",
                                                                       "YEAR",
                                                                       "visit_date",
                                                                       "ST_startdate",
                                                                       "ST_year1"
                                                                       )]

# Remove columns that have more than 11 NA values
pd_prodromal_df <- pd_prodromal_df[, !colSums(is.na(pd_prodromal_df)) > 11]
# Remove subjects that have NA values
pd_prodromal_df <- pd_prodromal_df[!rowSums(is.na(pd_prodromal_df)) > 0, ]
# Remove othneuro, tau_txt, and ptau_txt columns since  mostly "" characters
pd_prodromal_df <- pd_prodromal_df[!colnames(pd_prodromal_df) %in% c("othneuro", 
                                                                     "tau_txt",
                                                                     "ptau_txt")]
# Remove PD_MED_USE as all zeros
pd_prodromal_df <- pd_prodromal_df[!colnames(pd_prodromal_df) %in% c("PD_MED_USE")]
# Remove NP1DDS as it causes GLM not to converge; insignificant
pd_prodromal_df <- pd_prodromal_df[!colnames(pd_prodromal_df) %in% c("NP1DDS")]

# Make the patient number the rownames instead of own column
rownames(pd_prodromal_df) <- pd_prodromal_df$PATNO
pd_prodromal_df <- pd_prodromal_df[, !colnames(pd_prodromal_df) == "PATNO"]
# Make the cohort column the leftmost
pd_prodromal_df <- pd_prodromal_df[, c(ncol(pd_prodromal_df), 
                                       1:ncol(pd_prodromal_df)-1
                                       )] 
# Make the cohort column a factor to run GLM
pd_prodromal_df$cohort <- as.factor(pd_prodromal_df$cohort)

# Final dimensions are 712 patients by 78 variables (+ cohort variable)
# dim(pd_prodromal_df)


# Calculate p values from GLM for each column in relation to cohort 
p_values_by_column <- rep(NA, ncol(pd_prodromal_df))
for (i in 2:ncol(pd_prodromal_df)){
  p_values_by_column[i] <- coef(summary(glm(pd_prodromal_df$cohort ~ pd_prodromal_df[,i], 
                                            family = "binomial")))[2,4]
}

# Apply Bonferroni correction based on number of GLM tests ran
bonferroni_alpha <- 0.05/(ncol(pd_prodromal_df) - 1)
# Evaluate which columns are suggested to be signifcant
sig_factors <- colnames(pd_prodromal_df)[which(p_values_by_column < bonferroni_alpha)]

# Realization: remove variables that are simply categorical versions of others
# e.g. remove "age_cat", i.e., categorical variable for "age"
sig_factors_clean <- sig_factors[!sig_factors %in% c("age_cat",
                                                     "hy_on",
                                                     "NHY_ON",
                                                     "rigidity_on",
                                                     "td_pigd_on",
                                                     "tremor_on",
                                                     "updrs3_score_on",
                                                     "updrs_totscore_on",
                                                     "rem_cat"
                                                     )]

# Make dataframe with only significant columns
pd_prodromal_sig_df <- pd_prodromal_df[, colnames(pd_prodromal_df) %in% c("cohort", sig_factors_clean)]

# Run PCA
pca_results <- prcomp(pd_prodromal_sig_df[-1], center = TRUE, scale = TRUE)

# Plot PCA biplot with variables as directional lines
pca_biplot_fig <- fviz_pca_biplot(pca_results,
                                  col.ind = pd_prodromal_sig_df$cohort,
                                  addEllipses = TRUE,
                                  col.var = "black",
                                  label = "var",
                                  select.var = list(contrib = 10),
                                  repel = TRUE,
                                  labelsize = 3,
                                  legend.title = "Cohort") + 
  theme_classic() +
  ggtitle("PCA biplot from clinical data - PD vs. prodromal subjects") +
  theme(plot.title = element_text(hjust = 0.5))

pca_biplot_fig

# optionally save
#ggsave("figures/subplots/figS1_pca_biplot_prodromal_vs_pd.png", 
#       plot = pca_biplot_fig,
#       width = 1800,
#       height = 1500,
#       units = "px")
```


# Supplementary 2 - Tracking how sncRNAs correlate with motor symptoms in prodromal subjects over time

```{r, warning = FALSE}
# Boxplots of two significant transcripts identified in DE analysis
sncrna_norm_df_clean <- sncrna_norm_df[, c(PATNO_list %in% PATNO_list_clean)]
sncrna_norm_df_bl <- sncrna_norm_df_clean[, grepl("BL", colnames(sncrna_norm_df_clean))]
rownames(sncrna_norm_df_bl) <- sncrna_df$Reference
sncrna_norm_df_bl_t <- as.data.frame(t(sncrna_norm_df_bl))

pd_p_sig_transcripts_bl_df <- sncrna_norm_df_bl_t[, which(colnames(sncrna_norm_df_bl_t) %in% pd_p_sig_transcripts)]
pd_p_sig_transcripts_bl_df <- data.frame("norm_count" = c(as.numeric(pd_p_sig_transcripts_bl_df[,1]),
                                                          as.numeric(pd_p_sig_transcripts_bl_df[,2])),
                                         "transcript" = c(rep(colnames(pd_p_sig_transcripts_bl_df)[1], 
                                                            nrow(pd_p_sig_transcripts_bl_df)),
                                                          rep(colnames(pd_p_sig_transcripts_bl_df)[2], 
                                                            nrow(pd_p_sig_transcripts_bl_df))),
                                         "cohort" = rep(cohort_bl_list, 2))

pd_p_sig_transcripts_bl_df <- pd_p_sig_transcripts_bl_df[pd_p_sig_transcripts_bl_df$cohort != "SWEDD", ]
pd_p_sig_transcripts_bl_df$cohort <- factor(pd_p_sig_transcripts_bl_df$cohort, levels = c("Healthy", "Prodromal", "PD")) 

pd_p_sig_transcripts_bl_df_pval <- pd_p_sig_transcripts_bl_df %>% 
  rstatix::group_by(transcript) %>% 
  rstatix::t_test(norm_count ~ cohort) %>% 
  rstatix::add_xy_position()

pd_p_sig_transcripts_bl_df_pval$y.position <- c(3.6, 4.2, 3.9, 0.8, 1.4, 1.1)

pd_p_sig_transcripts_boxplots <-
  ggplot(
    pd_p_sig_transcripts_bl_df, 
    aes(x = cohort, y = log10(norm_count), fill = transcript)) + 
  geom_boxplot() +
  facet_wrap(~ transcript, scales = "free") + 
  xlab("") + 
  ylab("log10 normalized counts") +
  theme_prism() +
  add_pvalue(pd_p_sig_transcripts_bl_df_pval) + 
  guides(fill = FALSE) 

pd_p_sig_transcripts_boxplots

# optionally save
#ggsave("figures/subplots/figS2_pd_p_sig_transcripts_boxplots.png", 
#       plot = pd_p_sig_transcripts_boxplots,
#       width = 2500,
#       height = 1500,
#       units = "px")
```


```{r}
# Histogram of two significant sncRNAs identified in DE analysis
updrs_part3_df <- read.csv("Data/MDS-UPDRS_Part_III.csv")

updrs_part3_p_patno <- unique(updrs_part3_df$PATNO[which(updrs_part3_df$PATNO %in% p_patno_4plus_visits)])

updrs_part3_p <- list()
updrs_part3_p_change <- c()
updrs_part3_p_change_raw <- c()
for (i in 1:length(updrs_part3_p_patno)){
  updrs_part3_p[[i]] <- updrs_part3_df$NP3TOT[updrs_part3_df$PATNO == updrs_part3_p_patno[i]]
  totscore_start <- na.omit(updrs_part3_p[[i]])[1]
  totscore_end <- na.omit(updrs_part3_p[[i]])[length(na.omit(updrs_part3_p[[i]]))]
  updrs_part3_p_change_raw[i] <- totscore_end - totscore_start
  if (updrs_part3_p_change_raw[i] == 0){
    updrs_part3_p_change[i] <- 0 # no change
  } else if (updrs_part3_p_change_raw[i] > 0) {
    updrs_part3_p_change[i] <- -1 # worsened
  } else {
    updrs_part3_p_change[i] <- 1 # improved
  }
}

updrs_part3_p_patno_worsened <- updrs_part3_p_patno[which(updrs_part3_p_change == -1)]
updrs_part3_p_patno_improved <- updrs_part3_p_patno[which(updrs_part3_p_change == 1)]

visits_index <- list()
updrs_part3_p_with_tx_data <- list()
for (i in 1:length(updrs_part3_p_patno)){ # for each patient
  colnames_i <- colnames(sncrna_norm_df_p)[grepl(updrs_part3_p_patno[i], colnames(sncrna_norm_df_p))]
  visits_index[[i]]  <- substr(sub(".*.V", "", colnames_i), start = 1, stop = 2)
  for (j in 1:length(visits_index[[i]])){
    if (visits_index[[i]][j] == "PP"){
      visits_index[[i]][j] <- as.numeric(1)
    } else {
      visits_index[[i]][j] <- as.numeric(visits_index[[i]][j])
    }
  }
  visits_index[[i]] <- as.numeric(visits_index[[i]])
  updrs_part3_p_with_tx_data[[i]] <- nafill(updrs_part3_p[[i]][visits_index[[i]]], type = "locf")
  updrs_part3_p_with_tx_data[[i]] <- nafill(updrs_part3_p_with_tx_data[[i]], type = "nocb")
}

rho_p_all <- list()
for (i in 1:length(pd_p_sig_transcripts)){ # for each transcript
  rho_p_all[[i]] <- rep(NA, length(pd_p_sig_transcripts))

 for (j in 1:length(updrs_part3_p_patno)){ # for each patient
   sncrna_norm_df_p_i_j <- as.numeric(
     sncrna_norm_df_p[which(colnames(sncrna_norm_df_bl_t) %in% pd_p_sig_transcripts[i]),
                      grepl(updrs_part3_p_patno[j], 
                            colnames(sncrna_norm_df_p))]
     )
   
   updrs_part3_j <- updrs_part3_p_with_tx_data[[j]]
   rho_p_all[[i]][j] <- cor(sncrna_norm_df_p_i_j, updrs_part3_j, method = "spearman")
 } 
  # remove patients that did not exhibit any change in NPDRS score
  rho_p_all[[i]] <- na.omit(rho_p_all[[i]])
}

rho_p_all_df <- data.frame("rho" = c(rho_p_all[[1]],
                                     rho_p_all[[2]]),
                           "transcript" = c(rep(pd_p_sig_transcripts[1], length(rho_p_all[[1]])),
                                            rep(pd_p_sig_transcripts[2], length(rho_p_all[[1]]))
                                            ))

pd_p_sig_transcripts_rho_hist <-
  ggplot(rho_p_all_df,
         aes(x = rho, fill = transcript)) +
  geom_histogram(alpha = 0.4) +
  xlab("Spearman correlation between UPDRS part III\nmotor score and transcript expression") +
  ylab("Prodromal patient count") +
  geom_vline(xintercept =
               mean(rho_p_all_df$rho[rho_p_all_df$transcript == pd_p_sig_transcripts[1]]),
             color = "#F8766D",
             linewidth = 1.5,
             linetype = "longdash"
             ) +
  geom_vline(xintercept =
               mean(rho_p_all_df$rho[rho_p_all_df$transcript == pd_p_sig_transcripts[2]]),
             color = "#00C0B8",
             linewidth = 1.5,
             linetype = "longdash"
             ) +
  theme_prism()

pd_p_sig_transcripts_rho_hist

# optionally save
#ggsave("figures/subplots/figS2_pd_p_sig_transcripts_rho_hist.png", 
#       plot = pd_p_sig_transcripts_rho_hist,
#       width = 2500,
#       height = 1500,
#       units = "px")
```

```{r, warning = FALSE, message = FALSE}
# Distributions of top 4 sncRNAs that correlate with motor symptoms
# search all transcripts that have padj from DESeq2
pd_p_all_transcripts <- rownames(res_bl_pd_p_df)[which(!is.na(res_bl_pd_p_df$padj))]

mean_rho <- c()
rho_p_all <- list()
for (i in 1:length(pd_p_all_transcripts)){ # for each transcript
  rho_p_all[[i]] <- rep(NA, length(pd_p_all_transcripts))
 for (j in 1:length(updrs_part3_p_patno)){ # for each patient
   sncrna_norm_df_p_i_j <- as.numeric(
     sncrna_norm_df_p[which(colnames(sncrna_norm_df_bl_t) %in% pd_p_all_transcripts[i]),
                      grepl(updrs_part3_p_patno[j], 
                            colnames(sncrna_norm_df_p))]
     )
   
   updrs_part3_j <- updrs_part3_p_with_tx_data[[j]]
   rho_p_all[[i]][j] <- cor(sncrna_norm_df_p_i_j, updrs_part3_j, method = "spearman")
 } 
  # remove patients that did not exhibit any change in NPDRS score
  rho_p_all[[i]] <- na.omit(rho_p_all[[i]])
  mean_rho[i] <- mean(rho_p_all[[i]])
}

top10_indices_p <- order(abs(mean_rho), decreasing = TRUE)[1:10]
# 1 had too few samples; 3 and 5 had strange distribution shape
top4_indices_p_df <- data.frame(
  "transcript" = 
    c(rep(pd_p_all_transcripts[top10_indices_p[2]], length(rho_p_all[[top10_indices_p[2]]])),
      rep(pd_p_all_transcripts[top10_indices_p[4]], length(rho_p_all[[top10_indices_p[4]]])),
      rep(pd_p_all_transcripts[top10_indices_p[6]], length(rho_p_all[[top10_indices_p[6]]])),
      rep(pd_p_all_transcripts[top10_indices_p[7]], length(rho_p_all[[top10_indices_p[7]]]))
      ),
  "rho" = 
    c(rho_p_all[[top10_indices_p[2]]],
      rho_p_all[[top10_indices_p[4]]],
      rho_p_all[[top10_indices_p[6]]],
      rho_p_all[[top10_indices_p[7]]]
      ),
  "mean" = 
    c(rep(mean_rho[top10_indices_p[2]], length(rho_p_all[[top10_indices_p[2]]])),
      rep(mean_rho[top10_indices_p[4]], length(rho_p_all[[top10_indices_p[4]]])),
      rep(mean_rho[top10_indices_p[6]], length(rho_p_all[[top10_indices_p[6]]])),
      rep(mean_rho[top10_indices_p[7]], length(rho_p_all[[top10_indices_p[7]]]))
    )
)

top4_transcripts_p_density <-  
 ggplot(top4_indices_p_df,
        aes(x = rho, fill = transcript)) +
  geom_vline(aes(xintercept = top4_indices_p_df$mean,
                 color = top4_indices_p_df$transcript)
             ) + 
  geom_density(alpha = 0.4) +
  facet_wrap(~ transcript, scales = "free") +
  xlab("Spearman correlation between UPDRS part III\nmotor score and transcript expression") +
  ylab("Prodromal patient density") +
  guides(fill = FALSE) +
  guides(color = FALSE) +
  theme_prism()

top4_transcripts_p_density

# optionally save
#ggsave("figures/subplots/figS2_top4_transcripts_p_density.png", 
#       plot = top4_transcripts_p_density,
#       width = 2500,
#       height = 1500,
#       units = "px")
```


```{r, warning = FALSE, message = FALSE}
# Spin tests
mean_rho_spin <- c()
rho_p_all_spin <- list()
for (i in 1:length(pd_p_all_transcripts)){ # for each transcript
  rho_p_all_spin[[i]] <- rep(NA, length(pd_p_all_transcripts))
 for (j in 1:length(updrs_part3_p_patno)){ # for each patient
   updrs_part3_j <- updrs_part3_p_with_tx_data[[j]]
   sncrna_norm_df_p_i_j <- runif(length(updrs_part3_j), 0, 20)
   rho_p_all_spin[[i]][j] <- cor(sncrna_norm_df_p_i_j, updrs_part3_j, method = "spearman")
 } 
  # remove patients that did not exhibit any change in NPDRS score
  rho_p_all_spin[[i]] <- na.omit(rho_p_all_spin[[i]])
  mean_rho_spin[i] <- mean(rho_p_all_spin[[i]])
}

spin_p_df <- data.frame(
  "transcript" = 
    c(rep(pd_p_all_transcripts[top10_indices_p[2]], length(mean_rho_spin)),
      rep(pd_p_all_transcripts[top10_indices_p[4]], length(mean_rho_spin)),
      rep(pd_p_all_transcripts[top10_indices_p[6]], length(mean_rho_spin)),
      rep(pd_p_all_transcripts[top10_indices_p[7]], length(mean_rho_spin))
      ),
  "mean_rho_spin" = rep(mean_rho_spin, 4),
  "mean_rho_exp" = c(rep(top4_indices_p_df$mean[top4_indices_p_df$transcript == pd_p_all_transcripts[top10_indices_p[2]]][1], length(mean_rho_spin)),
                     rep(top4_indices_p_df$mean[top4_indices_p_df$transcript == pd_p_all_transcripts[top10_indices_p[4]]][1], length(mean_rho_spin)),
                     rep(top4_indices_p_df$mean[top4_indices_p_df$transcript == pd_p_all_transcripts[top10_indices_p[6]]][1], length(mean_rho_spin)),
                     rep(top4_indices_p_df$mean[top4_indices_p_df$transcript == pd_p_all_transcripts[top10_indices_p[7]]][1], length(mean_rho_spin))
                     )
  )


spin_top4_transcripts_p_density <-
  ggplot(spin_p_df,
            aes(x = mean_rho_spin)) +
  geom_vline(aes(xintercept = spin_p_df$mean_rho_exp,
                 color = spin_p_df$transcript)
             ) + 
  geom_histogram(aes(y=..density..), color="black", fill="white") +
  geom_density(alpha = .2, fill = "#E68613") +
  #facet_wrap(~ transcript, scales = "free") +
  xlab("Mean spearman correlation between UPDRS part III\nmotor score and transcript expression") +
  ylab("Prodromal patient density") +
  guides(fill = FALSE) +
  guides(color = FALSE) +
  ggtitle("Null dist. with top 4 from experimental dist. marked") + 
  theme_prism()

spin_top4_transcripts_p_density

# optionally save
#ggsave("figures/subplots/figS2_spin_top4_transcripts_p_density.png", 
#   plot = spin_top4_transcripts_p_density,
#   width = 2500,
#   height = 1500,
#   units = "px")
```


# Supplementary 3 - Tracking how sncRNAs correlate with motor symptoms in PD subjects over time


```{r, warning = FALSE}
# Histogram of five significant sncRNAs identified in DE analysis
sncrna_norm_df_clean <- sncrna_norm_df[, c(PATNO_list %in% PATNO_list_clean)]
sncrna_norm_df_bl <- sncrna_norm_df_clean[, grepl("BL", colnames(sncrna_norm_df_clean))]
rownames(sncrna_norm_df_bl) <- sncrna_df$Reference
sncrna_norm_df_bl_t <- as.data.frame(t(sncrna_norm_df_bl))

pd_h_sig_transcripts_bl_df <- sncrna_norm_df_bl_t[, which(colnames(sncrna_norm_df_bl_t) %in% pd_h_sig_transcripts)]
pd_h_sig_transcripts_bl_df <- data.frame("norm_count" = c(as.numeric(pd_h_sig_transcripts_bl_df[,1]),
                                                          as.numeric(pd_h_sig_transcripts_bl_df[,2]),
                                                          as.numeric(pd_h_sig_transcripts_bl_df[,3]),
                                                          as.numeric(pd_h_sig_transcripts_bl_df[,4]),
                                                          as.numeric(pd_h_sig_transcripts_bl_df[,5])),
                                         "transcript" = c(rep(colnames(pd_h_sig_transcripts_bl_df)[1], 
                                                            nrow(pd_h_sig_transcripts_bl_df)),
                                                          rep(colnames(pd_h_sig_transcripts_bl_df)[2], 
                                                            nrow(pd_h_sig_transcripts_bl_df)),
                                                          rep(colnames(pd_h_sig_transcripts_bl_df)[3], 
                                                            nrow(pd_h_sig_transcripts_bl_df)),
                                                          rep(colnames(pd_h_sig_transcripts_bl_df)[4], 
                                                            nrow(pd_h_sig_transcripts_bl_df)),
                                                          rep(colnames(pd_h_sig_transcripts_bl_df)[5], 
                                                            nrow(pd_h_sig_transcripts_bl_df))),
                                         "cohort" = rep(cohort_bl_list, 5))


pd_h_sig_transcripts_bl_df <- pd_h_sig_transcripts_bl_df[pd_h_sig_transcripts_bl_df$cohort != "SWEDD", ]
pd_h_sig_transcripts_bl_df$cohort[pd_h_sig_transcripts_bl_df$cohort == "Healthy"] <- "H"
pd_h_sig_transcripts_bl_df$cohort[pd_h_sig_transcripts_bl_df$cohort == "Prodromal"] <- "P"
pd_h_sig_transcripts_bl_df$cohort <- factor(pd_h_sig_transcripts_bl_df$cohort, levels = c("H", "P", "PD")) 

updrs_part3_df <- read.csv("../data/MDS-UPDRS_Part_III.csv")

sncrna_norm_df_pd <- sncrna_norm_df_clean[,which(sncrna_cohort_full_list == "PD")]

pd_patno <- unique(gsub("\\..*", "", substring(colnames(sncrna_norm_df_pd), 10)))

num_visits_p <- c()
for (i in 1:length(pd_patno)){
  num_visits_p[i] <- table(grepl(pd_patno[i], colnames(sncrna_norm_df_pd)))["TRUE"]
}

pd_patno_8plus_visits <- pd_patno[which(num_visits_p >= 8)]

sncrna_pd_8plus_df <- sncrna_df_clean[, grepl(paste(pd_patno_8plus_visits, collapse = "|"), colnames(sncrna_df_clean))]

updrs_part3_pd_patno <- unique(updrs_part3_df$PATNO[which(updrs_part3_df$PATNO %in% pd_patno_8plus_visits)])

updrs_part3_pd <- list()
for (i in 1:length(updrs_part3_pd_patno)){
  updrs_part3_pd[[i]] <- updrs_part3_df$NP3TOT[updrs_part3_df$PATNO == updrs_part3_pd_patno[i]]
}

visits_index <- list()
updrs_part3_pd_with_tx_data <- list()
for (i in 1:length(updrs_part3_pd_patno)){ # for each patient
  colnames_i <- colnames(sncrna_norm_df_pd)[grepl(updrs_part3_pd_patno[i], colnames(sncrna_norm_df_pd))]
  visits_index[[i]]  <- substr(sub(".*.V", "", colnames_i), start = 1, stop = 2)
  for (j in 1:length(visits_index[[i]])){
    if (visits_index[[i]][j] == "PP"){
      visits_index[[i]][j] <- as.numeric(1)
    } else {
      visits_index[[i]][j] <- as.numeric(visits_index[[i]][j])
    }
  }
  visits_index[[i]] <- as.numeric(visits_index[[i]])
  updrs_part3_pd_with_tx_data[[i]] <- nafill(updrs_part3_pd[[i]][visits_index[[i]]], type = "locf")
  updrs_part3_pd_with_tx_data[[i]] <- nafill(updrs_part3_pd_with_tx_data[[i]], type = "nocb")
}

rho_pd_all <- list()
for (i in 1:length(pd_h_sig_transcripts)){ # for each transcript
  rho_pd_all[[i]] <- rep(NA, length(pd_h_sig_transcripts))

 for (j in 1:length(updrs_part3_pd_patno)){ # for each patient
   sncrna_norm_df_pd_i_j <- as.numeric(
     sncrna_norm_df_pd[which(colnames(sncrna_norm_df_bl_t) %in% pd_h_sig_transcripts[i]),
                      grepl(updrs_part3_pd_patno[j], 
                            colnames(sncrna_norm_df_pd))]
     )
   
   updrs_part3_j <- updrs_part3_pd_with_tx_data[[j]]
   rho_pd_all[[i]][j] <- cor(sncrna_norm_df_pd_i_j, updrs_part3_j, method = "spearman")
 } 
  # remove patients that did not exhibit any change in NPDRS score
  rho_pd_all[[i]] <- na.omit(rho_pd_all[[i]])
}

rho_pd_all_df <- data.frame("rho" = c(rho_pd_all[[1]],
                                      rho_pd_all[[2]],
                                      rho_pd_all[[3]],
                                      rho_pd_all[[4]],
                                      rho_pd_all[[5]]),
                           "transcript" = c(rep(pd_h_sig_transcripts[1], length(rho_pd_all[[1]])),
                                            rep(pd_h_sig_transcripts[2], length(rho_pd_all[[1]])),
                                            rep(pd_h_sig_transcripts[3], length(rho_pd_all[[1]])),
                                            rep(pd_h_sig_transcripts[4], length(rho_pd_all[[1]])),
                                            rep(pd_h_sig_transcripts[5], length(rho_pd_all[[1]]))
                                            ))

pd_h_sig_transcripts_rho_hist <-
  ggplot(rho_pd_all_df,
         aes(x = rho, fill = transcript)) +
  geom_histogram(alpha = 0.6) +
  xlab("Spearman correlation between UPDRS part III\nmotor score and transcript expression") +
  ylab("PD patient count") +
  geom_vline(xintercept =
               mean(rho_pd_all_df$rho[rho_pd_all_df$transcript == pd_h_sig_transcripts[1]]),
             color = "#F8766D",
             linewidth = 1.5,
             linetype = "longdash"
             ) +
  geom_vline(xintercept =
               mean(rho_pd_all_df$rho[rho_pd_all_df$transcript == pd_h_sig_transcripts[2]]),
             color = "#00C0B8",
             linewidth = 1.5,
             linetype = "longdash"
             ) +
  theme_prism()

pd_h_sig_transcripts_rho_hist

# optionally save
#ggsave("figures/subplots/figS3_pd_h_sig_transcripts_rho_hist.png", 
#       plot = pd_h_sig_transcripts_rho_hist,
#       width = 2500,
#       height = 1500,
#       units = "px")
```

## Figure 6B - Top 4 sncRNAs that correlate with motor symptoms

```{r, warning = FALSE, message = FALSE}
# Distributions of top 4 sncRNAs that correlate with motor symptoms
# search all transcripts that have padj from DESeq2
pd_h_all_transcripts <- sncrna_df$Reference[which(!is.na(res_bl_pd_h_df$padj))]

mean_pd_rho <- c()
rho_pd_all <- list()
for (i in 1:length(pd_h_all_transcripts)){ # for each transcript
  rho_pd_all[[i]] <- rep(NA, length(pd_h_all_transcripts))
 for (j in 1:length(updrs_part3_pd_patno)){ # for each patient
   sncrna_norm_df_pd_i_j <- as.numeric(
     sncrna_norm_df_pd[which(colnames(sncrna_norm_df_bl_t) %in% pd_h_all_transcripts[i]),
                      grepl(updrs_part3_pd_patno[j], 
                            colnames(sncrna_norm_df_pd))]
     )
   rho_pd_all[[i]][j] <- cor(sncrna_norm_df_pd_i_j, updrs_part3_pd_with_tx_data[[j]], method = "spearman")
 } 
  # remove patients that did not exhibit any change in NPDRS score
  mean_pd_rho[i] <- mean(rho_pd_all[[i]], na.rm = TRUE)
}

top10_indices_pd <- order(abs(mean_pd_rho), decreasing = TRUE)[1:10]
# 1 had too few samples; 3 and 5 had strange distribution shape
top4_indices_pd_df <- data.frame(
  "transcript" = 
    c(rep(pd_h_all_transcripts[top10_indices_pd[2]], length(rho_pd_all[[top10_indices_pd[2]]])),
      rep(pd_h_all_transcripts[top10_indices_pd[4]], length(rho_pd_all[[top10_indices_pd[4]]])),
      rep(pd_h_all_transcripts[top10_indices_pd[6]], length(rho_pd_all[[top10_indices_pd[6]]])),
      rep(pd_h_all_transcripts[top10_indices_pd[7]], length(rho_pd_all[[top10_indices_pd[7]]]))
      ),
  "rho" = 
    c(rho_pd_all[[top10_indices_pd[2]]],
      rho_pd_all[[top10_indices_pd[4]]],
      rho_pd_all[[top10_indices_pd[6]]],
      rho_pd_all[[top10_indices_pd[7]]]
      ),
  "mean" = 
    c(rep(mean_pd_rho[top10_indices_pd[2]], length(rho_pd_all[[top10_indices_pd[2]]])),
      rep(mean_pd_rho[top10_indices_pd[4]], length(rho_pd_all[[top10_indices_pd[4]]])),
      rep(mean_pd_rho[top10_indices_pd[6]], length(rho_pd_all[[top10_indices_pd[6]]])),
      rep(mean_pd_rho[top10_indices_pd[7]], length(rho_pd_all[[top10_indices_pd[7]]]))
    )
)

top4_transcripts_pd_density <-  
 ggplot(top4_indices_pd_df,
        aes(x = rho, fill = transcript)) +
  geom_vline(aes(xintercept = top4_indices_pd_df$mean,
                 color = top4_indices_pd_df$transcript)
             ) + 
  geom_density(alpha = 0.4) +
  facet_wrap(~ transcript, scales = "free") +
  xlab("Spearman correlation between UPDRS part III\nmotor score and transcript expression") +
  ylab("PD patient density") +
  guides(fill = FALSE) +
  guides(color = FALSE) +
  theme_prism()

top4_transcripts_pd_density

# optionally save
#ggsave("figures/subplots/figS3_top4_transcripts_pd_density.png", 
#       plot = top4_transcripts_pd_density,
#       width = 2500,
#       height = 1500,
#       units = "px")
```

```{r, warning = FALSE, message = FALSE}
# Spin tests
mean_rho_pd_spin <- c()
rho_pd_all_spin <- list()
for (i in 1:length(pd_h_all_transcripts)){ # for each transcript
  rho_pd_all_spin[[i]] <- rep(NA, length(pd_h_all_transcripts))
 for (j in 1:length(updrs_part3_pd_patno)){ # for each patient
   updrs_part3_j <- updrs_part3_pd_with_tx_data[[j]]
   sncrna_norm_df_pd_i_j <- runif(length(updrs_part3_j), 0, 20)
   rho_pd_all_spin[[i]][j] <- cor(sncrna_norm_df_pd_i_j, updrs_part3_j, method = "spearman")
 } 
  # remove patients that did not exhibit any change in NPDRS score
  rho_pd_all_spin[[i]] <- na.omit(rho_pd_all_spin[[i]])
  mean_rho_pd_spin[i] <- mean(rho_pd_all_spin[[i]])
}

spin_pd_df <- data.frame(
  "transcript" = 
    c(rep(pd_h_all_transcripts[top10_indices_pd[2]], length(mean_rho_pd_spin)),
      rep(pd_h_all_transcripts[top10_indices_pd[4]], length(mean_rho_pd_spin)),
      rep(pd_h_all_transcripts[top10_indices_pd[6]], length(mean_rho_pd_spin)),
      rep(pd_h_all_transcripts[top10_indices_pd[7]], length(mean_rho_pd_spin))
      ),
  "mean_rho_pd_spin" = rep(mean_rho_pd_spin, 4),
  "mean_rho_pd_exp" = c(rep(top4_indices_pd_df$mean[top4_indices_pd_df$transcript == pd_h_all_transcripts[top10_indices_pd[2]]][1], length(mean_rho_pd_spin)),
                     rep(top4_indices_pd_df$mean[top4_indices_pd_df$transcript == pd_h_all_transcripts[top10_indices_pd[4]]][1], length(mean_rho_pd_spin)),
                     rep(top4_indices_pd_df$mean[top4_indices_pd_df$transcript == pd_h_all_transcripts[top10_indices_pd[6]]][1], length(mean_rho_pd_spin)),
                     rep(top4_indices_pd_df$mean[top4_indices_pd_df$transcript == pd_h_all_transcripts[top10_indices_pd[7]]][1], length(mean_rho_pd_spin))
                     )
  )


spin_top4_transcripts_pd_density <-
  ggplot(spin_pd_df,
            aes(x = mean_rho_pd_spin)) +
  geom_vline(aes(xintercept = spin_pd_df$mean_rho_pd_exp,
                 color = spin_pd_df$transcript)
             ) + 
  geom_histogram(aes(y=..density..), color="black", fill="white") +
  geom_density(alpha = .2, fill = "#E68613") +
  xlab("Mean spearman correlation between UPDRS part III\nmotor score and transcript expression") +
  ylab("PD patient density") +
  guides(fill = FALSE) +
  guides(color = FALSE) +
  ggtitle("Null dist. with top 4 from experimental dist. marked") + 
  theme_prism()

spin_top4_transcripts_pd_density
   
# optionally save
#ggsave("../figures/subplots/figS3_spin_top4_transcripts_pd_density.png", 
#   plot = spin_top4_transcripts_pd_density,
#   width = 2500,
#   height = 1500,
#   units = "px")
```
